src/core/thread/router_table.cpp

```diff
@@ -0,0 +1,634 @@
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define WPP_NAME "router_table.tmh"
+
+#include "router_table.hpp"
+
+#if OPENTHREAD_FTD
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/logging.hpp"
+#include "common/timer.hpp"
+#include "thread/mle.hpp"
+#include "thread/mle_router.hpp"
+#include "thread/network_data_leader.hpp"
+#include "thread/thread_netif.hpp"
+
+namespace ot {
+
+RouterTable::Iterator::Iterator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mRouter(NULL)
+{
+    Reset();
+}
+
+void RouterTable::Iterator::Reset(void)
+{
+    RouterTable &routerTable = GetInstance().Get<RouterTable>();
+
+    mRouter = &routerTable.mRouters[0];
+
+    if (mRouter->GetRloc16() == 0xffff)
+    {
+        mRouter = NULL;
+    }
+}
+
+void RouterTable::Iterator::Advance(void)
+{
+    RouterTable &routerTable = GetInstance().Get<RouterTable>();
+    Router *     listEnd     = &routerTable.mRouters[Mle::kMaxRouters];
+
+    VerifyOrExit(mRouter != NULL);
+
+    mRouter++;
+
+    VerifyOrExit(mRouter < listEnd && mRouter->GetRloc16() != 0xffff, mRouter = NULL);
+
+exit:
+    return;
+}
+
+RouterTable::RouterTable(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mRouterIdSequenceLastUpdated(0)
+    , mRouterIdSequence(Random::GetUint8())
+{
+    Clear();
+}
+
+void RouterTable::Clear(void)
+{
+    memset(mAllocatedRouterIds, 0, sizeof(mAllocatedRouterIds));
+    memset(mRouterIdReuseDelay, 0, sizeof(mRouterIdReuseDelay));
+    UpdateAllocation();
+}
+
+void RouterTable::ClearNeighbors(void)
+{
+    for (uint8_t i = 0; i < Mle::kMaxRouters; i++)
+    {
+        mRouters[i].SetState(Neighbor::kStateInvalid);
+    }
+}
+
+bool RouterTable::IsAllocated(uint8_t aRouterId) const
+{
+    return (mAllocatedRouterIds[aRouterId / 8] & (1 << (aRouterId % 8))) != 0;
+}
+
+void RouterTable::UpdateAllocation(void)
+{
+    uint8_t indexMap[Mle::kMaxRouterId + 1];
+    uint8_t routerCount = 0;
```

Comment:
Add `VerifyOrExit(mAllocatedRouterIds != 0)` here and put "// clear unused entries" after `exit` label?

Reply:
`mAllocatedRouterIds` is a byte array - or are you suggesting to first check if all the bytes are zero?

Reply:
...please ignore it.